### PR TITLE
Implements password hashing

### PR DESCRIPTION
**Pull request by Alkey**

### CODE REVIEW

**Review by Martynice — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java [hunk 1]
```diff
@@ -70,12 +71,13 @@ public User create(User user) {
...
         deleteUserRoles(user.getId());
-        String query = "UPDATE users SET name = ?, login = ?, password = ? "
+        String query = "UPDATE users SET name = ?, login = ?, password = ?, salt = ? "
                 + "WHERE user_id = ? AND deleted = false";
         try (PreparedStatement statement = ConnectionUtil.getConnection().prepareStatement(query)) {
             statement.setString(1, user.getName());
             statement.setString(2, user.getLogin());
             statement.setString(3, user.getPassword());
+            statement.setBytes(4, user.getSalt());
             statement.setLong(4, user.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/UserJdbcDaoImpl.java]:
```suggestion
            statement.setLong(5, user.getId());
```

File: src/main/java/com/internet/shop/dao/util/HashUtil.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(HASH_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalArgumentException("Couldn't hash password", e);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/util/HashUtil.java]:
```suggestion
            throw new RuntimeException("Couldn't hash password", e);
```
let's throw a more generic Exception - I think IllegalArgumentException does not suit here well

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -15,9 +16,13 @@
...
         User user = userService.getByLogin(login).orElseThrow(() ->
                 new AuthenticationException("Incorrect username"));
-        if (user.getPassword().equals(password)) {
+        if (isValid(user.getPassword(), password, user.getSalt())) {
             return user;
         }
         throw new AuthenticationException("Incorrect password");
     }
+
+    private static boolean isValid(String userPassword, String password, byte[] salt) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java]:
```suggestion
    private boolean isValid(String userPassword, String password, byte[] salt) {
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
let's also add `salt` column to init_db.sql

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**